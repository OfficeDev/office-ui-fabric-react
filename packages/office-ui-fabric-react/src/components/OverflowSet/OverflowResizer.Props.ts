import * as React from 'react';
import { IOverflowSetProps } from './OverflowSet.Props';

export interface IOverflowResizerProps extends React.HTMLProps<HTMLElement> {
  items?: any[];
  onOverflow?: (items: any[]) => { items: any[], overflow: any[] };
  overflowSetProps?: IOverflowSetProps;
}
