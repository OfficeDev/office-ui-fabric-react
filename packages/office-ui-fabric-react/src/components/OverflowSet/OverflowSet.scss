@import '../../common/common';

.root {
  position: relative;
  display: flex;
  flex-wrap: nowrap;
}

.item {
  flex-shrink: 0;
}

.overflowIcon {
  min-width: 0;
  padding: 0 4px;
  align-self: stretch;
  height: auto;
}
