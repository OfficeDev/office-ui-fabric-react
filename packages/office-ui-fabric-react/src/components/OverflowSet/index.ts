export * from './OverflowSet';
export * from './OverflowSet.Props';
export * from './OverflowResizer';
export * from './OverflowResizer.Props';
