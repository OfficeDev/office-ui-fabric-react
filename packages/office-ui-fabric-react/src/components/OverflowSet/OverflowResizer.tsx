import * as React from 'react';
import {
  css,
  autobind,
  BaseComponent,
  assign
} from '../../Utilities';
import { IOverflowResizerProps } from './OverflowResizer.Props';
import { OverflowSet } from './OverflowSet';
import { IContextualMenuProps } from '../../ContextualMenu';
import styles = require('./OverflowResizer.scss');

export interface IOverflowResizerState {
  renderedItems?: any[];
  renderedOverflowItems?: IContextualMenuProps[];
  measuredItems: any[];
  measuredOverflowItems: IContextualMenuProps[];
  shouldMeasure?: boolean;
}

export class OverflowResizer extends BaseComponent<IOverflowResizerProps, IOverflowResizerState> {

  private _root: HTMLElement;
  private _measured: HTMLElement;

  public static defaultProps = {
    items: []
  };

  constructor(props: IOverflowResizerProps) {
    super(props);
    this.state = {
      shouldMeasure: true,
      renderedItems: [],
      renderedOverflowItems: [],
      measuredItems: this.props.items.concat([]),
      measuredOverflowItems: []
    };
  }

  public componentDidMount() {
    this._measureItems();
    this._events.on(window, 'resize', this._onResize);
  }

  public render() {
    let { overflowSetProps } = this.props;
    let { shouldMeasure, renderedItems, renderedOverflowItems, measuredItems, measuredOverflowItems } = this.state;
    return (
      <div ref={ this._resolveRef('_root') }>
        { shouldMeasure && (
          <div className={ css(styles.measured) } ref={ this._resolveRef('_measured') }>
            <OverflowSet
              { ...assign(overflowSetProps, { items: measuredItems, overflowItems: measuredOverflowItems }) }
            />
          </div>
        ) }

        <OverflowSet
          { ...assign(overflowSetProps, { items: renderedItems, overflowItems: renderedOverflowItems }) }
        />
      </div>

    );
  }

  public componentDidUpdate(prevProps: IOverflowResizerProps) {
    this._measureItems();
  }

  protected _onResize() {
    this.setState({ shouldMeasure: true });
  }

  private _updatedOverflowState(prevState, props): IOverflowResizerState {
    let { measuredOverflowItems, measuredItems } = prevState;

    measuredOverflowItems.push(measuredItems.pop());
    return { measuredItems, measuredOverflowItems }
  }

  private _finalOverflowState(prevState, props): IOverflowResizerState {
    let { renderedItems, renderedOverflowItems, measuredItems, measuredOverflowItems } = prevState;
    return {
      renderedItems: measuredItems,
      renderedOverflowItems: measuredOverflowItems,
      measuredItems: this.props.items.concat([]),
      measuredOverflowItems: [],
      shouldMeasure: false
    };
  }

  private _measureItems() {
    let {
      items,
      updatedOverflowState = this._updatedOverflowState
    } = this.props;
    let {
      shouldMeasure,
      renderedItems,
      renderedOverflowItems,
      measuredItems,
      measuredOverflowItems
    } = this.state;

    if (shouldMeasure && items && items.length > 0) {
      let container = this._root.getBoundingClientRect();
      let measured = this._measured.getBoundingClientRect();
      if ((measured.width > container.width) ? true : false) {
        this.setState((prevState, props) => {
          return updatedOverflowState(prevState, props)
        });
      } else {
        this.setState((prevState, props) => {
          return assign({ shouldMeasure: false }, this._finalOverflowState(prevState, props));
        });
      }
    }
  }
}
