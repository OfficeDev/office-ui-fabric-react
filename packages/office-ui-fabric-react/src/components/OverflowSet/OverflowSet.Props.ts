import * as React from 'react';
import { OverflowSet } from './OverflowSet';
import { IContextualMenuItem } from '../../ContextualMenu';
import { IRenderFunction } from '../../Utilities';

export interface IOverflowSetProps extends React.Props<OverflowSet> {

  /**
   * An array of items to be rendered by your onRenderItem function in the primary content area
  */
  items: any[];

  /**
   * An array of items to be passed to overflow contextual menu
  */
  overflowItems?: IContextualMenuItem[];

  /**
   * Method to call when trying to render an item.
   * If item contains an onRender, that function will be used instead.
  */
  onRenderItem: IRenderFunction<any>;

  /**
   * Rendering method for overflow button and contextual menu.
  */
  onRenderOverflowButton: IRenderFunction<IContextualMenuItem[]>;
}
