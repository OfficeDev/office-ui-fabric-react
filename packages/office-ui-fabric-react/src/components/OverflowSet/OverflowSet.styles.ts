import { IOverflowSetProps, IOverflowSetStyleProps } from "./OverflowSet.types";

export const getStyles = (props: IOverflowSetProps): IOverflowSetStyleProps => {
  const { className, vertical } = props;
  return {
    root: [
      "ms-OverflowSet",
      {
        position: "relative",
        display: "flex",
        flexWrap: "nowrap"
      },
      vertical && { flexDirection: "column" },
      className
    ],
    item: {
      flexShrink: 0,
      display: "inherit"
    }
  };
};
