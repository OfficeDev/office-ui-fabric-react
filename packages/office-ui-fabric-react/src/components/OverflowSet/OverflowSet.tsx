import * as React from 'react';
import {
  css,
  BaseComponent,
  createRef,
  IKeySequence
} from '../../Utilities';
import { mergeStyles } from '../../Styling';
import { IOverflowSet, IOverflowSetProps, IOverflowSetItemProps } from './OverflowSet.types';
import { FocusZone, FocusZoneDirection } from '../../FocusZone';
import { KeytipManager } from '../../utilities/keytips';
import { IKeytipProps } from '../../Keytip';
import * as stylesImport from './OverflowSet.scss';

const styles: any = stylesImport;

export class OverflowSet extends BaseComponent<IOverflowSetProps, {}> implements IOverflowSet {

  private _focusZone = createRef<FocusZone>();
  private _persistedKeytips: IKeytipProps[] = [];
  private _keytipManager: KeytipManager = KeytipManager.getInstance();

  public render() {
    if (this._persistedKeytips.length) {
      // Delete all persisted keytips saved
      this._persistedKeytips.forEach((keytips: IKeytipProps) => {
        this._keytipManager.unregisterPersistedKeytip(keytips);
      });
      this._persistedKeytips = [];
    }

    const {
      items,
      overflowItems,
      className,
      focusZoneProps,
      vertical = false,
      role = 'menubar'
    } = this.props;

    return (
      <FocusZone
        { ...focusZoneProps }
        componentRef={ this._focusZone }
        className={ mergeStyles(
          'ms-OverflowSet',
          styles.root,
          vertical && styles.rootVertical,
          className
        ) }
        direction={ vertical ? FocusZoneDirection.vertical : FocusZoneDirection.horizontal }
        role={ role }
      >
        { items && this._onRenderItems(items) }
        { overflowItems && overflowItems.length > 0 && this._onRenderOverflowButtonWrapper(overflowItems) }
      </FocusZone>
    );
  }

  public focus() {
    if (this._focusZone.value) {
      this._focusZone.value.focus();
    }
  }

  public componentDidMount() {
    this._registerPersistedKeytips();
  }

  public componentDidUpdate() {
    this._registerPersistedKeytips();
  }

  private _registerPersistedKeytips() {
    this._persistedKeytips.forEach((keytips: IKeytipProps) => {
      this._keytipManager.registerPersistedKeytip(keytips);
    });
  }

  private _onRenderItems = (items: IOverflowSetItemProps[]): JSX.Element[] => {
    const overflowKeytipSequences = this.props.keytipSequences;

    return items.map((item, i) => {
      const wrapperDivProps: React.HTMLProps<HTMLDivElement> = { className: css('ms-OverflowSet-item', styles.item) };

      // Remove overflowSetSequence from all regular items
      if (overflowKeytipSequences && item.keytipProps) {
        this._removeOverflowSequence(item);
      }

      return (
        <div key={ item.key } { ...wrapperDivProps }>
          { this.props.onRenderItem(item) }
        </div>
      );
    });
  }

  private _onRenderOverflowButtonWrapper = (items: any[]): JSX.Element => {
    const wrapperDivProps: React.HTMLProps<HTMLDivElement> = { className: css('ms-OverflowSet-overflowButton', styles.item) };
    const overflowKeytipSequences = this.props.keytipSequences;

    // Register all persisted keytips and save
    if (overflowKeytipSequences) {
      items.forEach((overflowItem) => {
        const keytip = (overflowItem as IOverflowSetItemProps).keytipProps;
        if (keytip) {
          // Create persisted keytip
          const persistedKeytip: IKeytipProps = {
            content: keytip.content,
            keySequences: keytip.keySequences,
            hasChildrenNodes: false,
            disabled: keytip.disabled
          };

          if (keytip.hasChildrenNodes || overflowItem.subMenuProps) {
            // If the keytip has a submenu, change onExecute to persistedKeytipExecute
            persistedKeytip.onExecute = this._keytipManager.persistedKeytipExecute.bind(this._keytipManager, overflowKeytipSequences, overflowItem.keytipProps.keySequences);
          } else {
            // If the keytip doesn't have a submenu, just execute the original function
            persistedKeytip.onExecute = (el: HTMLElement | null) => {
              keytip.onExecute && keytip.onExecute(el);
            };
          }

          // Add this persisted keytip to our internal list
          this._persistedKeytips.push(persistedKeytip);

          // Add the overflow sequence to this item and its subMenu items
          this._addOverflowSequence(overflowKeytipSequences, overflowItem);
        }
      });
    }
    return (
      <div { ...wrapperDivProps }>
        { this.props.onRenderOverflowButton(items) }
      </div>
    );
  }

  /**
   * Unsets overflowSetSequence on item and its subMenu items
   */
  private _removeOverflowSequence(item: any): void {
    item.keytipProps.overflowSetSequence = undefined;

    if (item.subMenuProps) {
      item.subMenuProps.items.forEach((subMenuItem: any) => {
        if (subMenuItem.keytipProps) {
          this._removeOverflowSequence(subMenuItem);
        }
      });
    }
  }

  /**
   * Sets overflowSetSequence on item and its subMenu items
   */
  private _addOverflowSequence(overflowSequence: IKeySequence[], overflowItem: any): void {
    overflowItem.keytipProps.overflowSetSequence = overflowSequence;

    if (overflowItem.subMenuProps) {
      overflowItem.subMenuProps.items.forEach((subMenuItem: any) => {
        if (subMenuItem.keytipProps) {
          this._addOverflowSequence(overflowSequence, subMenuItem);
        }
      });
    }
  }

}