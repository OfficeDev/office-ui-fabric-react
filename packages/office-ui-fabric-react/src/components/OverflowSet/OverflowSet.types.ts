import * as React from 'react';
import { OverflowSet } from './OverflowSet';
import { IRenderFunction, IKeySequence } from '../../Utilities';
import { IFocusZoneProps } from '../../FocusZone';
import { IKeytipProps } from '../../Keytip';

export interface IOverflowSet {
  /**
 * Sets focus to the button.
 */
  focus: () => void;
}

export interface IOverflowSetProps extends React.Props<OverflowSet> {
  /**
   * Gets the component ref.
   */
  componentRef?: (ref?: IOverflowSet) => void;

  /**
   * Class name
   */
  className?: string;

  /**
   * An array of items to be rendered by your onRenderItem function in the primary content area
  */
  items?: IOverflowSetItemProps[];

  /**
   * Change item layout direction to vertical/stacked
   * @default false
  */
  vertical?: boolean;

  /**
   * An array of items to be passed to overflow contextual menu
  */
  overflowItems?: IOverflowSetItemProps[];

  /**
   * Method to call when trying to render an item.
  */
  onRenderItem: (item: IOverflowSetItemProps) => any;

  /**
   * Rendering method for overflow button and contextual menu. The argument to the function is
   * the overflowItems passed in as props to this function.
  */
  onRenderOverflowButton: IRenderFunction<any[]>;

  /**
   * Custom properties for OverflowSet's FocusZone.
   */
  focusZoneProps?: IFocusZoneProps;

  /**
   * The role for the OverflowSet.
   * @default 'menubar'
   */
  role?: string;

  /**
   * Optional keytip sequence for the overflow button, if it will have a keytip
   */
  keytipSequences?: IKeySequence[];
}

export interface IOverflowSetItemProps {
  /**
   * Unique id to identify the item
   */
  key: string;

  /**
   * Optional keytip for the overflowSetItem
   */
  keytipProps?: IKeytipProps;

  /**
   * Any additional properties to use when custom rendering menu items.
   */
  [propertyName: string]: any;
}