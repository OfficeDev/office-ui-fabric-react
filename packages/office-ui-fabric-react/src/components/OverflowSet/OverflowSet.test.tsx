import * as React from 'react';
import { shallow } from 'enzyme';

import { OverflowSet } from './OverflowSet';
import * as sinon from 'sinon';
import { IOverflowSetItemProps } from './OverflowSet.types';
import { CommandBarButton } from '../../Button';
import { mount, ReactWrapper } from 'enzyme';
import { convertSequencesToKeytipID, fullKeySequencesAreEqual, IKeySequence } from '../../Utilities';
import { IKeytipProps } from '../../Keytip';
import { KeytipManager, constructKeytipExecuteTargetFromId, KeytipTree, IKeytipTreeNode } from '../../utilities/keytips';

function getTreeNode(keytipTree: KeytipTree, keySequences: IKeySequence[]): IKeytipTreeNode | undefined {
  return keytipTree.getNode(convertSequencesToKeytipID(keySequences));
}

describe('OverflowSet', () => {
  it('does not render overflow when there are no overflow items', () => {
    const onRenderItem = sinon.spy();
    const onRenderOverflowButton = sinon.spy();
    shallow(<OverflowSet onRenderItem={ onRenderItem } onRenderOverflowButton={ onRenderOverflowButton } />);

    expect(onRenderOverflowButton.called).toEqual(false);
  });

  it('does not render overflow when overflow items is an empty array', () => {
    const onRenderItem = sinon.spy();
    const onRenderOverflowButton = sinon.spy();
    shallow(<OverflowSet onRenderItem={ onRenderItem } onRenderOverflowButton={ onRenderOverflowButton } overflowItems={ [] } />);

    expect(onRenderOverflowButton.called).toEqual(false);
  });

  function delay(millisecond: number): Promise<void> {
    return new Promise<void>((resolve) => setTimeout(resolve, millisecond));
  }

  describe('keytip tests', () => {
    let overflowSet: ReactWrapper;
    let overflowKeytips: any;
    let items: IOverflowSetItemProps[];
    let overflowItems: IOverflowSetItemProps[];

    const onRenderItem = (item: IOverflowSetItemProps): JSX.Element => {
      return (
        <CommandBarButton
          { ...item }
          menuProps={ item.subMenuProps }
        >{ item.name }
        </CommandBarButton>
      );
    };

    const onRenderOverflowButton = (overflowElements: any[] | undefined): JSX.Element => {
      return (
        <CommandBarButton
          menuIconProps={ { iconName: 'More' } }
          menuProps={ { items: overflowElements! } }
          keytipProps={ overflowKeytips.overflowButtonKeytip }
        />
      );
    };

    const keytipManager = KeytipManager.getInstance();

    beforeEach(() => {
      keytipManager.currentSequence = '';

      overflowKeytips = {
        overflowItemKeytip1: {
          content: 'A',
          keySequences: ['a'],
          onExecute: jest.fn()
        },
        overflowItemKeytip2: {
          content: 'B',
          keySequences: ['b'],
          onExecute: jest.fn()

        },
        overflowItemKeytip3: {
          content: 'C',
          keySequences: ['c'],
          onExecute: jest.fn()

        },
        overflowItemKeytip4: {
          content: 'D',
          keySequences: ['d'],
          onExecute: jest.fn()
        },
        overflowItemKeytip5: {
          content: 'E',
          keySequences: ['d', 'e'],
          onExecute: jest.fn()
        },
        overflowItemKeytip6: {
          content: 'F',
          keySequences: ['d', 'f'],
          onExecute: jest.fn()
        },
        overflowButtonKeytip: {
          // Overflow button
          content: 'X',
          keySequences: ['x'],
          onExecute: (el: HTMLElement) => {
            // Find the overflow button and manually click it to open the overflow menu
            overflowSet.find(constructKeytipExecuteTargetFromId('ktp-x')).simulate('click');
          }
        }
      };

      items = [
        {
          key: 'item1',
          name: 'Item 1',
          keytipProps: overflowKeytips.overflowItemKeytip1
        },
        {
          key: 'item2',
          name: 'Item 2',
          keytipProps: overflowKeytips.overflowItemKeytip2
        },
      ];

      overflowItems = [
        {
          key: 'item3',
          name: 'Item 3',
          keytipProps: overflowKeytips.overflowItemKeytip3
        },
        {
          key: 'item4',
          name: 'Item 4',
          keytipProps: overflowKeytips.overflowItemKeytip4
        },
      ];

      overflowSet = mount(
        (
          <OverflowSet
            onRenderItem={ onRenderItem }
            onRenderOverflowButton={ onRenderOverflowButton }
            items={ items }
            keytipSequences={ overflowKeytips.overflowButtonKeytip.keySequences }
          />
        ));
    });

    afterEach(() => {
      // Clean up the keytip items
      keytipManager.keytips = [];
      keytipManager.keytipTree = new KeytipTree();

      // Cleanup ContextualMenus that were created
      for (let i = 0; i < document.body.children.length; i++) {
        if (document.body.children[i].tagName === 'DIV') {
          document.body.removeChild(document.body.children[i]);
          i--;
        }
      }
    });

    describe('without submenus', () => {

      beforeEach(() => {
        overflowSet.setProps({
          overflowItems
        });
      });

      it('should register regular and persisted keytips', () => {
        // Persisted keytips will have the original key sequence of the items in the overflow
        const keytipTree = keytipManager.keytipTree;
        // Regular keytips
        expect(getTreeNode(keytipTree, overflowKeytips.overflowItemKeytip1.keySequences)).toBeDefined();
        expect(getTreeNode(keytipTree, overflowKeytips.overflowItemKeytip2.keySequences)).toBeDefined();
        // Persisted keytips
        expect(getTreeNode(keytipTree, overflowKeytips.overflowItemKeytip3.keySequences)).toBeDefined();
        expect(getTreeNode(keytipTree, overflowKeytips.overflowItemKeytip4.keySequences)).toBeDefined();
        // Overflow button keytip
        expect(getTreeNode(keytipTree, overflowKeytips.overflowButtonKeytip.keySequences)).toBeDefined();
      });

      it('should properly register and unregister keytips when items are moved to the overflow and back', () => {
        // Add the first overflow item to 'items'
        overflowSet.setProps({
          items: items.concat(overflowItems.slice(0, 1)),
          overflowItems: overflowItems.slice(1, 2)
        });

        // Should still have 5 keytips in the tree
        const keytipTree = keytipManager.keytipTree;
        // Regular keytips
        expect(getTreeNode(keytipTree, overflowKeytips.overflowItemKeytip1.keySequences)).toBeDefined();
        expect(getTreeNode(keytipTree, overflowKeytips.overflowItemKeytip2.keySequences)).toBeDefined();
        expect(getTreeNode(keytipTree, overflowKeytips.overflowItemKeytip3.keySequences)).toBeDefined();
        // Persisted keytips
        expect(getTreeNode(keytipTree, overflowKeytips.overflowItemKeytip4.keySequences)).toBeDefined();
        // Overflow button keytip
        expect(getTreeNode(keytipTree, overflowKeytips.overflowButtonKeytip.keySequences)).toBeDefined();
      });

      it('persisted keytip should execute function when triggered', () => {
        // Set current keytip at root, like we've entered keytip mode
        keytipManager.keytipTree.currentKeytip = keytipManager.keytipTree.root;
        keytipManager.processInput('d');

        // D will be a persisted keytip, but it should call the onExecute of D
        expect(overflowKeytips.overflowItemKeytip4.onExecute).toBeCalled();

        // D has no children, we should be out of keytip mode (no current keytip)
        expect(keytipManager.keytipTree.currentKeytip).toBeUndefined();

        // No keytips should be visible
        keytipManager.getKeytips().forEach((keytip: IKeytipProps) => {
          expect(keytip.visible).toBeFalsy();
        });
      });

      it('triggering the overflow button keytip should register the menu item keytips with their modified sequence', () => {
        // Set current keytip at root, like we've entered keytip mode
        const keytipTree = keytipManager.keytipTree;
        keytipTree.currentKeytip = keytipTree.root;
        // Open the overflow menu
        keytipManager.processInput('x');

        delay(750).then(() => {
          // Opening the submenu should register the two keytips for those items
          const modifiedKeytip3Sequence = ['x', 'c'];
          const modifiedKeytip4Sequence = ['x', 'd'];
          expect(getTreeNode(keytipTree, modifiedKeytip3Sequence)).toBeDefined();
          expect(getTreeNode(keytipTree, modifiedKeytip4Sequence)).toBeDefined();

          // Those two keytips should now be visible in the Layer
          const submenuKeytips = keytipManager.getKeytips().filter((keytip: IKeytipProps) => {
            return keytip.content === 'C' || keytip.content === 'D';
          });
          submenuKeytips.forEach((submenuKeytip: IKeytipProps) => {
            expect(submenuKeytip.visible).toEqual(true);
          });
        });
      });

      it('overflowSetSequence gets set and unset correctly on overflowItems keytipProps', () => {
        // overflowSetSequence should be set on item3 and item4
        let item3OverflowSequence = overflowItems[0].keytipProps!.overflowSetSequence;
        expect(fullKeySequencesAreEqual(item3OverflowSequence!, overflowKeytips.overflowButtonKeytip.keySequences)).toEqual(true);
        let item4OverflowSequence = overflowItems[1].keytipProps!.overflowSetSequence;
        expect(fullKeySequencesAreEqual(item4OverflowSequence!, overflowKeytips.overflowButtonKeytip.keySequences)).toEqual(true);

        // Add the first overflow item to 'items'
        overflowSet.setProps({
          items: items.concat(overflowItems.slice(0, 1)),
          overflowItems: overflowItems.slice(1, 2)
        });

        // overflowSetSequence should not be set on item3, but should be set on item4
        item3OverflowSequence = overflowItems[0].keytipProps!.overflowSetSequence;
        expect(item3OverflowSequence).toBeUndefined();
        item4OverflowSequence = overflowItems[1].keytipProps!.overflowSetSequence;
        expect(fullKeySequencesAreEqual(item4OverflowSequence!, overflowKeytips.overflowButtonKeytip.keySequences)).toEqual(true);
      });
    });

    describe('with submenus', () => {
      let item3: any;
      beforeEach(() => {
        item3 = { ...overflowItems[0] };
      });

      describe('without children keytips', () => {
        it('should not exit keytip mode after being triggered', () => {
          // Insert a submenu into one of the overflow items
          const overflowItemsWithSubMenu = [
            item3,
            {
              key: 'item4',
              name: 'Item 4',
              keytipProps: {
                ...overflowKeytips.overflowItemKeytip4,
                onExecute: (el: HTMLElement) => {
                  el.click();
                }
              },
              subMenuProps: {
                items: [
                  {
                    key: 'item5',
                    name: 'Item 5'
                  },
                  {
                    key: 'item6',
                    name: 'Item 6'
                  },
                ]
              }
            },
          ];

          overflowSet.setProps({
            overflowItems: overflowItemsWithSubMenu
          });

          keytipManager.keytipTree.currentKeytip = keytipManager.keytipTree.root;
          keytipManager.processInput('d');

          expect(keytipManager.keytipTree.currentKeytip).toBeDefined();
        });
      });

      describe('with children keytips', () => {
        let overflowItemsWithSubMenuAndKeytips: any[];

        beforeEach(() => {
          overflowItemsWithSubMenuAndKeytips = [
            item3,
            {
              key: 'item4',
              name: 'Item 4',
              keytipProps: {
                ...overflowKeytips.overflowItemKeytip4,
                onExecute: (el: HTMLElement) => {
                  el.click();
                }
              },
              subMenuProps: {
                items: [
                  {
                    key: 'item5',
                    name: 'Item 5',
                    keytipProps: overflowKeytips.overflowItemKeytip5
                  },
                  {
                    key: 'item6',
                    name: 'Item 6',
                    keytipProps: overflowKeytips.overflowItemKeytip6,
                    subMenuProps: {
                      items: [
                        {
                          key: 'item7',
                          name: 'Item 7',
                          keytipProps: {
                            content: 'X',
                            keySequences: [{ keys: ['d'] }, { keys: ['f'] }, { keys: ['x'] }]
                          }
                        }
                      ]
                    }
                  },
                ]
              }
            },
          ];

          overflowSet.setProps({
            overflowItems: overflowItemsWithSubMenuAndKeytips
          });
        });

        it('subMenuProps overflowSetSequence get set and unset correctly', () => {
          // subMenuProps overflowSetSequence should be set on item5, item6, and item7
          // item5
          let subMenuProps = (overflowItemsWithSubMenuAndKeytips[1] as any).subMenuProps.items;
          expect(fullKeySequencesAreEqual(subMenuProps[0].keytipProps.overflowSetSequence,
            overflowKeytips.overflowButtonKeytip.keySequences)).toEqual(true);
          // item6
          expect(fullKeySequencesAreEqual(subMenuProps[1].keytipProps.overflowSetSequence,
            overflowKeytips.overflowButtonKeytip.keySequences)).toEqual(true);
          // item7
          let item6SubMenu = subMenuProps[1].subMenuProps!.items as any;
          expect(fullKeySequencesAreEqual(item6SubMenu[0].keytipProps.overflowSetSequence,
            overflowKeytips.overflowButtonKeytip.keySequences)).toEqual(true);

          // Move item4 to items
          overflowSet.setProps({
            items: items.concat(overflowItemsWithSubMenuAndKeytips.slice(1, 2)),
            overflowItems: overflowItemsWithSubMenuAndKeytips.slice(0, 1)
          });

          // subMenuProps overflowSetSequence should be unset on item5, item6, and item7
          // item5
          subMenuProps = (overflowItemsWithSubMenuAndKeytips[1] as any).subMenuProps!.items;
          expect(subMenuProps[0].keytipProps.overflowSetSequence).toBeUndefined();
          // item6
          expect(subMenuProps[1].keytipProps.overflowSetSequence).toBeUndefined();
          // item7
          item6SubMenu = subMenuProps[1].subMenuProps!.items as any;
          expect(item6SubMenu[0].keytipProps.overflowSetSequence).toBeUndefined();
        });

        it('should open the overflow and submenu when the persisted keytip is triggered', () => {
          const keytipTree = keytipManager.keytipTree;
          keytipTree.currentKeytip = keytipTree.root;
          keytipManager.processInput('d');

          // The two submenu keytips should be registered with their modified sequence in the tree
          const modifiedKeytip5Sequence = ['x', 'd', 'e'];
          const modifiedKeytip6Sequence = ['x', 'd', 'f'];
          const subMenu5Keytip = getTreeNode(keytipTree, modifiedKeytip5Sequence);
          expect(subMenu5Keytip).toBeDefined();
          const subMenu6Keytip = getTreeNode(keytipTree, modifiedKeytip6Sequence);
          expect(subMenu6Keytip).toBeDefined();

          delay(750).then(() => {
            // Those two keytips should now be visible in the Layer
            const submenuKeytips = keytipManager.getKeytips().filter((keytip: IKeytipProps) => {
              return keytip.content === 'E' || keytip.content === 'F';
            });
            submenuKeytips.forEach((submenuKeytip: IKeytipProps) => {
              expect(submenuKeytip.visible).toEqual(true);
            });
          });
        });
      });

      describe('with non-standard children keytips', () => {
        it('should respect itemSubMenuProvider when setting overflowSetSequence', () => {
          const overflowItemsWithSubMenuAndKeytips = [
            item3,
            {
              key: 'item4',
              name: 'Item 4',
              keytipProps: {
                ...overflowKeytips.overflowItemKeytip4,
                onExecute: (el: HTMLElement) => {
                  el.click();
                }
              },
              customSubMenu: {
                items: [
                  {
                    key: 'item5',
                    name: 'Item 5',
                    keytipProps: overflowKeytips.overflowItemKeytip5
                  },
                  {
                    key: 'item6',
                    name: 'Item 6',
                    keytipProps: overflowKeytips.overflowItemKeytip6,
                    customSubMenu: {
                      items: [
                        {
                          key: 'item7',
                          name: 'Item 7',
                          keytipProps: {
                            content: 'X',
                            keySequences: [{ keys: ['d'] }, { keys: ['f'] }, { keys: ['x'] }]
                          }
                        }
                      ]
                    }
                  },
                ]
              }
            },
          ];

          overflowSet.setProps({
            overflowItems: overflowItemsWithSubMenuAndKeytips,
            itemSubMenuProvider: (item: IOverflowSetItemProps) => {
              if (item.customSubMenu) {
                return item.customSubMenu.items;
              }
              return undefined;
            }
          });

          // Verify that the overflow set sequence is set correctly on item5, item6, and item7
          // item5
          const customSubMenu = (overflowItemsWithSubMenuAndKeytips[1] as any).customSubMenu.items;
          expect(fullKeySequencesAreEqual(customSubMenu[0].keytipProps.overflowSetSequence,
            overflowKeytips.overflowButtonKeytip.keySequences)).toEqual(true);
          // item6
          expect(fullKeySequencesAreEqual(customSubMenu[1].keytipProps.overflowSetSequence,
            overflowKeytips.overflowButtonKeytip.keySequences)).toEqual(true);
          // item7
          const item6SubMenu = customSubMenu[1].customSubMenu!.items as any;
          expect(fullKeySequencesAreEqual(item6SubMenu[0].keytipProps.overflowSetSequence,
            overflowKeytips.overflowButtonKeytip.keySequences)).toEqual(true);
        });
      });
    });
  });
});