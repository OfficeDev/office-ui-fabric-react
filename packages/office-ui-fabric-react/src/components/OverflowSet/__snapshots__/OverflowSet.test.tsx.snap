// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OverflowSet snapshot tests basicSnapshot 1`] = `
<div
  aria-describedby={undefined}
  aria-labelledby={undefined}
  className="ms-FocusZone ms-OverflowSet"
  data-focuszone-id="FocusZone0"
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseDownCapture={[Function]}
  role="menubar"
/>
`;

exports[`OverflowSet snapshot tests snapshot with classname 1`] = `
<div
  aria-describedby={undefined}
  aria-labelledby={undefined}
  className="ms-FocusZone ms-OverflowSet foobar"
  data-focuszone-id="FocusZone1"
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseDownCapture={[Function]}
  role="menubar"
/>
`;

exports[`OverflowSet snapshot tests snapshot with classname and vertical layout 1`] = `
<div
  aria-describedby={undefined}
  aria-labelledby={undefined}
  className="ms-FocusZone ms-OverflowSet foobar"
  data-focuszone-id="FocusZone2"
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseDownCapture={[Function]}
  role="menubar"
/>
`;
